docs/learn/bossroom/architecture.md

```diff
@@ -0,0 +1,266 @@
+
+DI also allows Boss Room to circumvent the problem of cross-scene references to common dependencies, even though it still has to manage the lifecycle of MonoBehaviour-based dependencies by marking them with `DontDestroyOnLoad` and destroying them manually when appropriate.
+
+:::note
+`ApplicationController` inherits from the `VContainer`'s `LifetimeScope`, a class that serves as a dependency injection scope and bootstrapper that facilitates binding dependencies. Scene-specific State classes inherit from `LifetimeScope`, too.
+:::
+
+In the Unity Editor Inspector, you can choose a parent scope for any `LifetimeScope`. When doing so, it’s helpful to set a cross-scene reference to some parent scopes, most commonly the `ApplicationController`. Setting a cross-scene reference allows you to bind scene-specific dependencies while maintaining easy access to the global dependencies of the `ApplicationController` in the State-specific version of a `LifetimeScope` object.
+
+### `Client`/`Server` code separation
+
+A challenge we encountered when developing Boss Room was that code would often run in a single context, either client or server. Reading mixed client and server code adds a layer of complexity, making mistakes more likely.
+
+To solve this challenge, we explored different client-server code separation approaches. We eventually decided to revert our initial client/server/shared assemblies to a more classic domain-driven assembly architecture while keeping more complex classes separated by client/server.
+
+Our initial thought was that separating assemblies by client and server would allow for easier porting to a dedicated game server (DGS) model afterward; we’d only need to strip a single assembly to ensure code only runs when necessary.
+
+However, this approach had two primary issues:
+
+* It introduced callback hell, which made code (that should be trivial) too complex. You can look at our different action implementations in Boss Room 1.3.1 to see this.
+* Many components could be single simple classes instead of three-class horrors.
+
+After investigation, we determined that client/server separation was unnecessary for the following reasons:
+
+* There’s no need for [asmdef (assembly definition)](https://docs.unity3d.com/Manual/cus-asmdef.html) stripping because you can ifdef out single classes instead.
+* It’s not completely necessary to ifdef classes because it’s only compile-time insurance that certain parts of client-side code never run.
```

Comment:
```suggestion
* It’s not completely necessary to ifdef classes because it’s only compile-time insurance that certain parts of client-side code never run. You can still disable the component on Awake at runtime if it's not mean to run on the server or client.
```